youtube_dl/extractor/picosong.py

```diff
@@ -0,0 +1,57 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from ..utils import urljoin
+
+
+class PicosongIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?picosong\.com/(?P<id>[0-9A-Za-z]{4})'
+    _TESTS = [{
+        'url': 'http://picosong.com/7uer/',
+        'md5': '7f2ee011ba3ca2ac180a35deb10feabf',
+        'info_dict': {
+            'id': '7uer',
+            'ext': 'mp3',
+            'artist': 'Xarlable',
+            'title': 'Rock My Forum',
+            'album': 'Epic',
+            'genre': 'Howto & Style',
+        },
+    }, {
+        'url': 'http://picosong.com/i3gs/',
+        'md5': 'd4ec5d64d97e28b45d2710c096c29f70',
+        'info_dict': {
+            'id': 'i3gs',
+            'ext': 'mp3',
+            'title': 'Rick.mp3',
+        },
+    }]
+
+    def _real_extract(self, url):
+        def extract_info_td(key, fatal=True):
+            return self._html_search_regex(
+                r'%s:</td>\s*<td class="value">([^<]*)</td>' % (key.capitalize(), ),
+                page, key, fatal=fatal)
+
+        video_id = self._match_id(url)
+
+        page = self._download_webpage(url, video_id)
+        # <param name="movie" value="/static/swf/picoplayer.swf?file=/cdn/085a4492f79c0c487bb656afa902213e.mp3"/>
+        download_path = self._search_regex(
+            r'value="/static/swf/picoplayer.swf\?file=([^"]+)"',
+            page, 'url')
+        download_url = urljoin(url, download_path)
+
+        title = extract_info_td('title')
+        if not title:
+            title = extract_info_td('filename')
```

Comment:
Oh, OK. Sorry, I misunderstood you. That's not what the filename fallback is about. There are three possible scenarios here:

1. Title is present:
```
<td class="key">Title:</td>
<td class="value">Rock My Forum</td>
```

This is the case of the first testcase. In this case, we want to use "Rock My Forum" as the title here.

2. The title field is seen but blank:
```
<td class="key">Title:</td>
<td class="value"></td>
```

In this case, I chose to fall back to the filename. This is the case of the second testcase ("Rick.mp3") and is what this code is about.

3. The fields are not on the web page at all. In this case, this likely means that picosong.com is down or broken, so this is a fatal error, which should work as intended.

This is what my code is trying to do. Does this help clear this code up?